### PR TITLE
explicitly throw exception if PropertyChange event is abstract

### LINKED ISSUE #473: Crash upon abstract PropertyChanged event declaration
**TL;DR**: Having PropertyChanged event declared as abstract causes NRE in `ModuleWeaver.GetEventHandlerField()` due to `addMethods.Single().Body.Instructions` being null.
___

**Long story**: here's the minimal setup to reproduce the bug:

``` C#
    abstract class BaseClass : INotifyPropertyChanged
    {
        public abstract event PropertyChangedEventHandler PropertyChanged;
    }
```

I understand that such setup is very uncommon, and I've even run into it by mistake. But then I've spent a few hours debugging the issue, because the actual crash doesn't reveal much of useful info:

```
1>MSBUILD : error : Fody: An unhandled exception occurred:
1>MSBUILD : error : Exception:
1>MSBUILD : error : Failed to execute weaver C:\Users\user\source\repos\FodyTest\packages\PropertyChanged.Fody.3.1.3\build\..\weaver\PropertyChanged.Fody.dll
1>MSBUILD : error : Type:
1>MSBUILD : error : System.Exception
1>MSBUILD : error : StackTrace:
1>MSBUILD : error :    at InnerWeaver.ExecuteWeavers() in C:\projects\fody\FodyIsolated\InnerWeaver.cs:line 210
1>MSBUILD : error :    at InnerWeaver.Execute() in C:\projects\fody\FodyIsolated\InnerWeaver.cs:line 111
1>MSBUILD : error : Source:
1>MSBUILD : error : FodyIsolated
1>MSBUILD : error : TargetSite:
1>MSBUILD : error : Void ExecuteWeavers()
1>MSBUILD : error : Object reference not set to an instance of an object.
1>MSBUILD : error : Type:
1>MSBUILD : error : System.NullReferenceException
1>MSBUILD : error : StackTrace:
1>MSBUILD : error :    at ModuleWeaver.GetEventHandlerField(TypeDefinition targetType)
1>MSBUILD : error :    at ModuleWeaver.InjectMethod(TypeDefinition targetType, InvokerTypes& invokerType)
1>MSBUILD : error :    at ModuleWeaver.AddOnPropertyChangedMethod(TypeDefinition targetType)
1>MSBUILD : error :    at ModuleWeaver.FindMethodsForNodes()
1>MSBUILD : error :    at ModuleWeaver.Execute()
1>MSBUILD : error :    at InnerWeaver.ExecuteWeavers() in C:\projects\fody\FodyIsolated\InnerWeaver.cs:line 198
1>MSBUILD : error : Source:
1>MSBUILD : error : PropertyChanged.Fody
1>MSBUILD : error : TargetSite:
1>MSBUILD : error : Mono.Cecil.FieldDefinition GetEventHandlerField(Mono.Cecil.TypeDefinition)
1>MSBUILD : error : 
```
You see, no mention of problematic type. So I think at least proper exception from `ModuleWeaver.GetEventHandlerField()` like the one below could be useful:

``` C#
    if (addMethods.Single().IsAbstract)
        throw new ArgumentException($"{targetType}.PropertyChanged is abstract");
```

If somebody would kindly give a clue which way the issue should be fixed, I'm all on it. :)

### PR DESCRIPTION
**Pull request by grayed**
#### Description

(copied from #473)

**TL;DR**: Having PropertyChanged event declared as abstract causes NRE in `ModuleWeaver.GetEventHandlerField()` due to `addMethods.Single().Body.Instructions` being null.
___

**Long story**: here's the minimal setup to reproduce the bug:

``` C#
    abstract class BaseClass : INotifyPropertyChanged
    {
        public abstract event PropertyChangedEventHandler PropertyChanged;
    }
```

I understand that such setup is very uncommon, and I've even run into it by mistake. But then I've spent a few hours debugging the issue, because the actual crash doesn't reveal much of useful info:

```
1>MSBUILD : error : Fody: An unhandled exception occurred:
1>MSBUILD : error : Exception:
1>MSBUILD : error : Failed to execute weaver C:\Users\user\source\repos\FodyTest\packages\PropertyChanged.Fody.3.1.3\build\..\weaver\PropertyChanged.Fody.dll
1>MSBUILD : error : Type:
1>MSBUILD : error : System.Exception
1>MSBUILD : error : StackTrace:
1>MSBUILD : error :    at InnerWeaver.ExecuteWeavers() in C:\projects\fody\FodyIsolated\InnerWeaver.cs:line 210
1>MSBUILD : error :    at InnerWeaver.Execute() in C:\projects\fody\FodyIsolated\InnerWeaver.cs:line 111
1>MSBUILD : error : Source:
1>MSBUILD : error : FodyIsolated
1>MSBUILD : error : TargetSite:
1>MSBUILD : error : Void ExecuteWeavers()
1>MSBUILD : error : Object reference not set to an instance of an object.
1>MSBUILD : error : Type:
1>MSBUILD : error : System.NullReferenceException
1>MSBUILD : error : StackTrace:
1>MSBUILD : error :    at ModuleWeaver.GetEventHandlerField(TypeDefinition targetType)
1>MSBUILD : error :    at ModuleWeaver.InjectMethod(TypeDefinition targetType, InvokerTypes& invokerType)
1>MSBUILD : error :    at ModuleWeaver.AddOnPropertyChangedMethod(TypeDefinition targetType)
1>MSBUILD : error :    at ModuleWeaver.FindMethodsForNodes()
1>MSBUILD : error :    at ModuleWeaver.Execute()
1>MSBUILD : error :    at InnerWeaver.ExecuteWeavers() in C:\projects\fody\FodyIsolated\InnerWeaver.cs:line 198
1>MSBUILD : error : Source:
1>MSBUILD : error : PropertyChanged.Fody
1>MSBUILD : error : TargetSite:
1>MSBUILD : error : Mono.Cecil.FieldDefinition GetEventHandlerField(Mono.Cecil.TypeDefinition)
1>MSBUILD : error : 
```
You see, no mention of problematic type. So I think at least proper exception from `ModuleWeaver.GetEventHandlerField()` like the one below could be useful:

``` C#
    if (addMethods.Single().IsAbstract)
        throw new ArgumentException($"{targetType}.PropertyChanged is abstract");
```

#### The solution

That's a very uncommon and doubtful case, so no reason to put a work on real support. But generating proper exception will be useful.

#### Todos

fixes #473 
is ever a test needed?
no documentation changes required

### CODE REVIEW

**Comment by grayed** [comment]
> LGTM. A test would be nice though - it would require a separate test assembly since it's an error case.

Hm. You get an exception in any case: either NRE, or WeaverException. so the only thing test could do is to check whether it is a WeaverException, or something else. Is this what you mean?

Thinking more, it looks like that putting some try/catch-throw proactively at the places particular types/members get manipulated would be useful. I.e.:

``` C#
var foo = LookupFoo(bar);
if (foo is null)
    throw new WeaverException($"foo not found in {bar.FullName}");
try
{
    // do something with foo
}
catch (WeaverException we) { throw; }    // already handled
catch (Exception e) {
    throw new WeaverException($"error while handling {foo.FullName}", e);
}
```
This way user would be able to easily localize the problem and, becoming able to file proper PR and/or tweak the code that causes problem.

In this case a new unit test could check for exceptions being thrown, that they all are WeaverExceptions, and that they all contain problematic type/member name.

Does this sound like a plan? I think I would be able to cook something like that in a few days, the PropertyChanged.Fody codebase isn't that large. :)

**Comment by ltrzesniewski** [comment]
That sounds good IMO, but I'd ask @SimonCropp what he thinks about this idea beforehand.

**Comment by SimonCropp** [comment]
i think just keep this PR scoped to the current fix. 

the `error while handling {foo.FullName}` all over the place is not as useful as u suspect given since the majority of IL issues are exposed at run time, assembly load time, or by using PEVerify.

If we think the "error while handling" is worth the effort i would prefer to look into approaches where the fody core can do most of the boilerplate code

**Comment by grayed** [comment]
@SimonCropp I just thought that such explicit exceptions would be a low hanging fruit. Of course you know what's better, so I stop my verbiage now. :)

Then what should I do, regarding this small issue which we talk a lot of already? :) Write the unit test the @ltrzesniewski said about... testing the type of exception being thrown? Or what?

**Comment by ltrzesniewski** [comment]
Maybe it's not really necessary to test this after all. Do we merge this as-is @SimonCropp?

**Comment by SimonCropp** [comment]
i am happy as is. if someone wants  to go to the effort of test i am happy with that as well

**Comment by ltrzesniewski** [comment]
OK, let's merge then. Thanks @grayed!

**Comment by SimonCropp** [comment]
This is now deployed. NuGet may take some time to make it available for download.

thanks @grayed and @ltrzesniewski
